test/controllers/GroupSchemeControllerSpec.scala

```diff
@@ -133,99 +139,248 @@ class GroupSchemeControllerSpec
-      when(mockSessionService.fetchAll()(any())).thenReturn(Future.successful(cacheItem))
-      when(mockSessionService.cache(any(), any[CompanyDetailsList]())(any(), any()))
-        .thenReturn(Future(sessionPair))
+      status(result) shouldBe BAD_REQUEST
 
-      val result = testGroupSchemeController.showDeleteCompany(0)(authRequest)
+      val document = Jsoup.parse(contentAsString(result))
+      document.getElementsByTag("h1").text shouldBe Messages("ers.csop.group_confirm_delete_company.page_header", Fixtures.companyName)
+      document.getElementsByClass("govuk-error-summary__title").text shouldBe Messages("validation.summary.heading")
+      document.getElementById("value-error").text shouldBe "Error: " + Messages("ers.csop.group_confirm_delete_company.err.message")
+    }
+
+    "give a redirect to groupSchemePage if the 'No' radio option is selected" in {
+      setAuthMocks()
+      val authRequest =
+        buildRequestWithAuth(Fixtures.buildFakeRequestWithSessionIdCSOP("POST").withFormUrlEncodedBody(("value", "false")))
+      when(mockSessionService.fetchCompaniesOptionally()(any(), any())).thenReturn(Future.successful(companyDetailsList))
+      when(mockSessionService.fetch[RequestObject](refEq(mockErsUtil.ERS_REQUEST_OBJECT))(any(), any()))
+        .thenReturn(Future.successful(ersRequestObject.copy(
+          schemeName = Some("CSOP"),
+          schemeType = Some("CSOP")
+        )))
+
+      val result = testGroupSchemeController.confirmDeleteCompanySubmit(0)(authRequest)
 
       status(result) shouldBe SEE_OTHER
+      headers(result)(implicitly)("Location") shouldBe "/submit-your-ers-annual-return/subsidiary-company-summary"
+    }
 
-      verify(mockSessionService, times(1)).cache(refEq(SUBSIDIARY_COMPANIES_CACHE), refEq(expected))(any(), any())
+    "give a redirect to getGlobalErrorPage if the company is fails to delete" in {
+      setAuthMocks()
+      val authRequest =
+        buildRequestWithAuth(Fixtures.buildFakeRequestWithSessionIdCSOP("POST").withFormUrlEncodedBody(("value", "true")))
+      when(mockSessionService.fetchCompaniesOptionally()(any(), any())).thenReturn(Future.successful(companyDetailsList))
+      when(mockSessionService.fetch[RequestObject](refEq(mockErsUtil.ERS_REQUEST_OBJECT))(any(), any()))
+        .thenReturn(Future.successful(ersRequestObject.copy(
+          schemeName = Some("CSOP"),
+          schemeType = Some("CSOP")
+        )))
+      when(mockCompanyDetailsService.deleteCompany(any())(any())).thenReturn(Future.successful(false))
+
+      val result = testGroupSchemeController.confirmDeleteCompanySubmit(0)(authRequest)
+
+      status(result) shouldBe INTERNAL_SERVER_ERROR
+
+      val document = Jsoup.parse(contentAsString(result))
+      document.getElementsByTag("title").get(0).text shouldBe Messages("ers.global_errors.title")
     }
-  }
 
+    "give a redirect to getGlobalErrorPage if the final company fails to delete" in {
```

Comment:
The test for multiple companies and a single company are basically the same test you can combine them like this:
```scala annotate
    Seq(("a single", companyDetailsListSingle), ("multiple", companyDetailsListSingle)).map{
      case (numberCompanies, companyDetails) =>
        s"give a redirect to getGlobalErrorPage if $numberCompanies company is fails to delete" in {
            setAuthMocks()
            val authRequest = ....

```